Jetpack Redirector: Add backend support for redirecting inbound links to their proper destination instead of 404ing

jblz opened the pull request:
This is a Work in progress! Do not merge :)

See p1HpG7-58b-p2

---

#### Changes proposed in this Pull Request:

* Register a new `Jetpack_Redirect` Custom Post Type with no UI
* Attempt to resolve 404 requests to such a post type
  * Redirect on a hit
  * Do nothing on a miss
* Cache results to speed up subsequent requests

#### Testing instructions:

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.

Would you like this feature to be tested by Beta testers as well?
Please add instructions to to-test.md in a new commit as part of your PR.
-->

* Determine the post ID of the intended target (it can be any type of post including regular posts / pages, attachment pages, or custom post types). It will be referred to as `$target_post_id` below.
* Manually add a redirect (via `wp shell` or anywhere else you can run a one-off command):
```
// This can be pretty much any arbitrary path
$path_to_redirect = '/Your/way/KEWL/old/path/you/really/want/to/redirect';

wp_insert_post( array(
    'post_type' => 'jetpack-redirect',
    'post_status' => 'publish',
    'post_parent' => $target_post_id,
    'post_title' => $path_to_redirect,
    'post_name' => md5( $path_to_redirect ),
) );
```
* Attempt to fetch the path you used above.  For example:
`curl -I https://random-gibberish.jurassic.ninja/Your/way/KEWL/old/path/you/really/want/to/redirect`

You should see the path resolve to the intended content via a `301` response with custom header `x-jetpack-redirector: HIT`

It behave appropriately in all supported versions of PHP (~5.2+)

<!-- Add the following only if this is meant to be in changelog -->
#### Proposed changelog entry for your changes:

jblz committed:
Legacy Redirector: first pass at the class for the backend

jblz committed:
hook in the added class

marekhrabe commented:
I suggest removing the "legacy" word completely. This feature is very universal and could be just as useful for people migrating legacy links as it might be for people that just want to make any kind of redirect URL.

zinigor commented:
Sorry for spamming the issue with matticbot comments, I promise not to overdo it :)

jblz commented on _inc/lib/class.jetpack-legacy-redirector.php:
```diff
@@ -0,0 +1,191 @@
+				FROM $wpdb->posts
+				WHERE post_type = %s
+				AND post_name = %s
+				LIMIT 1",
```
Should we also do `ADD post_status = 'publish'`?

jblz committed:
Move to `modules/custom-post-types` and load in `module-extras.php`

jblz committed:
remove an indent level

jblz committed:
remove some needless `elseif`s

jblz committed:
Support redirecting directly to attached files...
...if the extension is allowed & matches

jblz commented on modules/custom-post-types/redirector.php:
```diff
@@ -0,0 +1,211 @@
+				ID
+				FROM $wpdb->posts
+				WHERE post_type = %s
+				AND post_name = %s
```
Should we also do `ADD post_status = 'publish'`?

marekhrabe replied:
Great suggestion!

jblz commented on modules/custom-post-types/redirector.php:
```diff
@@ -0,0 +1,211 @@
+	static function get_redirect_post_id( $url ) {
+		global $wpdb;
+
+		$redirect_post_id = $wpdb->get_var( $wpdb->prepare(
```
Should we use a `WP_Query` here instead? It would add some amount caching, but this query is indexed & super fast as-is.

jblz committed:
Add unit tests & a few normalization enhancements

jblz committed:
Set `rewrite` to false in cpt registration
Rewrites are not needed & it was causing test failures.
Also, no need to specify `'public' => false` -- it's the default.

jblz committed:
Remove initialization tests

jblz committed:
Add a test for get_url_hash

marekhrabe commented on modules/custom-post-types/redirector.php:
```diff
@@ -0,0 +1,220 @@
+ */
+
+class Jetpack_Redirector {
+	const POST_TYPE = 'jetpack-redirect';
```
The core seems to prefer underscored post types https://codex.wordpress.org/Post_Types

Should Jetpack follow or not?

jblz replied:
Good thought, but I'm following the existing pattern from `jetpack-portfolio`, etc.:

https://github.com/Automattic/jetpack/blob/de978b5e46f04d4586f9026578c95a219c0cb86c/modules/custom-post-types/portfolios.php#L4

jblz committed:
use WP_Query instead of wpdb & add `publish` constraint

jblz committed:
use md5 -- we cannot use sha224 in PHP5.2 :(

jblz committed:
Don't compute the # of rows found

jblz committed:
* add a filter to disable the feature entirely
* bail from the redirect function when `is_admin`
* sanitize and unslash all the input
* bail if `wp_validate_redirect` doesn't pass

jblz commented on modules/module-extras.php:
```diff
@@ -16,6 +16,7 @@
 	'theme-tools/social-menu.php',
 	'theme-tools/content-options.php',
 	'custom-post-types/comics.php',
+	'custom-post-types/redirector.php',
```
Placed here, the tool is only `init`ed when Jetpack has a connection or it's in development mode. I'm pretty sure this is the correct behavior here, but we can also include it below to run whenever the plugin is active (regardless of "connected" status). 

@marekhrabe @jeherve do you have opinions here?

jeherve replied:
This should be fine imo.

jblz commented:
Closing as this has gotten too stale. New initiative => new PR.
